doc/doxygen/man/man3/Catalogues.3
.TH "Catalogues" 3 "Wed Oct 6 2021" "MAGRATHEA/PATHFINDER" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Catalogues \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <catalogues\&.h>\fP
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "template<class Parameters , class Map > static void \fBReadParamFile\fP (Parameters &\fBparameters\fP, Map &parameter)"
.br
.RI "\fIRead parameter file\&. \fP"
.ti -1c
.RI "template<typename Integer , class Parameter > static void \fBReadParticlesHDF5\fP (const Integer rank, const Parameter &\fBparameters\fP, std::vector< std::array< double, 8 > > &caractVect_source)"
.br
.RI "\fIRead Particles from HDF5 files\&. \fP"
.ti -1c
.RI "template<typename Integer , class Parameter > static void \fBReadParticlesASCII\fP (const Integer rank, const Parameter &\fBparameters\fP, std::vector< std::array< double, 8 > > &caractVect_source)"
.br
.RI "\fIRead Particles from ASCII files\&. \fP"
.ti -1c
.RI "template<int Order = ORDER, bool RK4 = true, bool Verbose = false, class Point , class Cosmology , class Octree , class Type , class Parameter > static std::array< std::array
.br
< double, 2 >, 2 > \fBnewtonMethod2d\fP (const Point &vobs, const Point &observer, const Point &trueTarget, const std::array< double, 2 > &target, const Point &velocity, std::array< std::array< double, 2 >, 2 > &jacobian, const Parameter &\fBparameters\fP, const Cosmology &cosmology, const Octree &octree, const Type length, const Type h, std::vector< double > &redshifts, double &interpRef, const unsigned int iteration)"
.br
.RI "\fIIterate once using Newton Method\&. \fP"
.ti -1c
.RI "template<int Order = ORDER, bool RK4 = true, bool Verbose = false, class Point , class Cosmology , class Octree , class Type , class Parameter > static std::array< std::array
.br
< double, 2 >, 2 > \fBiterateNewtonMethod\fP (const Point &vobs, const Point &observer, const Type phi, const Type theta, const Point &target, const Point &velocity, std::array< std::array< double, 2 >, 2 > &jacobian, const Parameter &\fBparameters\fP, const Cosmology &cosmology, const Octree &octree, const Type length, const Type h, std::vector< double > &redshifts, double &interpRef)"
.br
.RI "\fIIterate Newton Method\&. \fP"
.ti -1c
.RI "template<class Point , class Cosmology , class Octree , class Type , class Parameter > static void \fBrelCat\fP (const Point &vobs, const std::array< std::array< double, 3 >, 3 > &rotm1, std::string &nomOutput, const Point &observer, const std::vector< std::array< double, 8 > > &targets_position, const std::vector< std::array< double, 18 > > &previous_catalogue, const Parameter &\fBparameters\fP, const Cosmology &cosmology, const Octree &octree, const Type length, const Type h)"
.br
.RI "\fIGive new position of points\&. \fP"
.ti -1c
.RI "template<class Point , class Cosmology , class Octree , class Type , class Parameter > static void \fBrelCat_with_previous_cat\fP (const Point &vobs, std::string &nomOutput, const Point &observer, std::vector< std::array< double, 18 > > &previous_catalogue, const Parameter &\fBparameters\fP, const Cosmology &cosmology, const Octree &octree, const Type length, const Type h)"
.br
.RI "\fIGive new position of points\&. \fP"
.in -1c
.SH "Member Function Documentation"
.PP 
.SS "template<int Order, bool RK4, bool Verbose, class Point , class Cosmology , class Octree , class Type , class Parameter > std::array< std::array< double, 2 >, 2 > Catalogues::iterateNewtonMethod (const Point &vobs, const Point &observer, const Typephi, const Typetheta, const Point &target, const Point &velocity, std::array< std::array< double, 2 >, 2 > &jacobian, const Parameter &parameters, const Cosmology &cosmology, const Octree &octree, const Typelength, const Typeh, std::vector< double > &redshifts, double &interpRef)\fC [static]\fP"

.PP
Iterate Newton Method\&. Iterate Newton Method to a certain precision 
.PP
\fBTemplate Parameters:\fP
.RS 4
\fIOrder\fP Octree interpolation order : 0 for NGP, 1 for CIC, 2 for TSC or -1 for a homogeneous universe\&. 
.br
\fIRK4\fP Runge-kutta of fourth order or euler\&. 
.br
\fIVerbose\fP Verbose mode for debug purposes\&. 
.br
\fIPoint\fP point type\&. 
.br
\fICosmology\fP Cosmology evolution type\&. 
.br
\fIOctree\fP Octree type\&. 
.br
\fIType\fP Scalar type\&. 
.br
\fIParameter\fP Parameter type\&. 
.RE
.PP
\fBParameters:\fP
.RS 4
\fIvobs\fP Observer peculiar velocity, in SI 
.br
\fIobserver\fP Observer position\&. 
.br
\fIphi\fP Source comoving angular coordinate phi\&. 
.br
\fItheta\fP Source comoving angular coordinate theta\&. 
.br
\fItarget\fP Comoving position of the source 
.br
\fIvelocity\fP Velocity of the target\&. 
.br
\fIjacobian\fP Jacobian matrix 
.br
\fIparameters\fP Parameter structure 
.br
\fIcosmology\fP Cosmology evolution\&. 
.br
\fIoctree\fP Octree\&. 
.br
\fIlength\fP Spatial length in SI units\&. 
.br
\fIh\fP Dimensionless Hubble parameter 
.br
\fIredshifts\fP Vector containing the redshift decomposition for a given source\&. 
.br
\fIinterpRef\fP value for interpolation of the bundle\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
2x2 array with NEW initial angles given after several iterations of newton method and errors in angles at same radius for the NEW initial angles\&. 
.RE
.PP

.SS "template<int Order, bool RK4, bool Verbose, class Point , class Cosmology , class Octree , class Type , class Parameter > std::array< std::array< double, 2 >, 2 > Catalogues::newtonMethod2d (const Point &vobs, const Point &observer, const Point &trueTarget, const std::array< double, 2 > &target, const Point &velocity, std::array< std::array< double, 2 >, 2 > &jacobian, const Parameter &parameters, const Cosmology &cosmology, const Octree &octree, const Typelength, const Typeh, std::vector< double > &redshifts, double &interpRef, const unsigned intiteration)\fC [static]\fP"

.PP
Iterate once using Newton Method\&. Fit the better initial conditions for the photon that intersects the source\&. 
.PP
\fBTemplate Parameters:\fP
.RS 4
\fIOrder\fP Octree interpolation order : 0 for NGP, 1 for CIC, 2 for TSC or -1 for an homogeneous universe\&. 
.br
\fIRK4\fP Runge-kutta of fourth order or euler\&. 
.br
\fIVerbose\fP Verbose mode for debug purposes\&. 
.br
\fIPoint\fP point type\&. 
.br
\fICosmology\fP Cosmology evolution type\&. 
.br
\fIOctree\fP Octree type\&. 
.br
\fIType\fP Scalar type\&. 
.br
\fIParameter\fP Parameter type 
.RE
.PP
\fBParameters:\fP
.RS 4
\fIvobs\fP Observer peculiar velocity, in SI 
.br
\fIobserver\fP Observer cartesian position\&. 
.br
\fItrueTarget\fP True cartesian target position\&. 
.br
\fItarget\fP Observed angular target position\&. 
.br
\fIvelocity\fP Target velocity\&. 
.br
\fIjacobian\fP Jacobian matrix 
.br
\fIparameters\fP Parameters structure 
.br
\fIcosmology\fP Cosmology evolution\&. 
.br
\fIoctree\fP Octree\&. 
.br
\fIlength\fP Spatial length in SI units\&. 
.br
\fIh\fP Dimensionless Hubble parameter 
.br
\fIredshifts\fP Vector containing the redshift decomposition for a given source\&. 
.br
\fIinterpRef\fP value for interpolation of the bundle\&. 
.br
\fIiteration\fP Number of iterations for the root-finder 
.RE
.PP
\fBReturns:\fP
.RS 4
2x2 array with NEW initial angles given by newton method and angle difference at the source between source and photon 
.RE
.PP

.SS "template<class Parameters , class Map > void Catalogues::ReadParamFile (Parameters &parameters, Map &parameter)\fC [static]\fP"

.PP
Read parameter file\&. Read and put in a structure the parameters\&. 
.PP
\fBTemplate Parameters:\fP
.RS 4
\fIParameters\fP structure type 
.br
\fIMap\fP map type 
.RE
.PP
\fBParameters:\fP
.RS 4
\fIparameters\fP Structure containing the parameters\&. 
.br
\fIparameter\fP Contains parameters to be rewritten 
.RE
.PP
\fBReturns:\fP
.RS 4
Filled parameters structure\&. 
.RE
.PP

.SS "template<typename Integer , class Parameter > void Catalogues::ReadParticlesASCII (const Integerrank, const Parameter &parameters, std::vector< std::array< double, 8 > > &caractVect_source)\fC [static]\fP"

.PP
Read Particles from ASCII files\&. Read particles from ASCII files and put them in vectors 
.PP
\fBTemplate Parameters:\fP
.RS 4
\fIInteger\fP Integer type 
.br
\fIParameter\fP Parameter type 
.RE
.PP
\fBParameters:\fP
.RS 4
\fIrank\fP Rank 
.br
\fIparameters\fP Parameters structure 
.br
\fIcaractVect_source\fP Source caracteristics 
.RE
.PP

.SS "template<typename Integer , class Parameter > void Catalogues::ReadParticlesHDF5 (const Integerrank, const Parameter &parameters, std::vector< std::array< double, 8 > > &caractVect_source)\fC [static]\fP"

.PP
Read Particles from HDF5 files\&. Read particles from HDF5 files and put them in vectors 
.PP
\fBTemplate Parameters:\fP
.RS 4
\fIInteger\fP Integer type 
.br
\fIParameter\fP Parameter type 
.RE
.PP
\fBParameters:\fP
.RS 4
\fIrank\fP Rank 
.br
\fIparameters\fP Parameters structure 
.br
\fIcaractVect_source\fP Source caracteristics 
.RE
.PP

.SS "template<class Point , class Cosmology , class Octree , class Type , class Parameter > void Catalogues::relCat (const Point &vobs, const std::array< std::array< double, 3 >, 3 > &rotm1, std::string &filename, const Point &observer, const std::vector< std::array< double, 8 > > &targets_position, const std::vector< std::array< double, 18 > > &previous_catalogue, const Parameter &parameters, const Cosmology &cosmology, const Octree &octree, const Typelength, const Typeh)\fC [static]\fP"

.PP
Give new position of points\&. Give lensed position of points given points\&. 
.PP
\fBTemplate Parameters:\fP
.RS 4
\fIPoint\fP point type\&. 
.br
\fICosmology\fP Cosmology evolution type\&. 
.br
\fIOctree\fP Octree type\&. 
.br
\fIType\fP Scalar type\&. 
.br
\fIParameter\fP Parameter type\&. 
.RE
.PP
\fBParameters:\fP
.RS 4
\fIvobs\fP Observer peculiar velocity, in SI 
.br
\fIrotm1\fP rotation matrix 
.br
\fIfilename\fP output file name\&. 
.br
\fIobserver\fP Observer position\&. 
.br
\fItargets_position\fP vector of target positions, velocities and radius\&. 
.br
\fIprevious_catalogue\fP Previously computed catalogue, used to re-run rejected sources\&. 
.br
\fIparameters\fP Parameter structure 
.br
\fIcosmology\fP Cosmology evolution\&. 
.br
\fIoctree\fP Octree\&. 
.br
\fIlength\fP Spatial length in SI units\&. 
.br
\fIh\fP Dimensionless Hubble parameter 
.RE
.PP

.SS "template<class Point , class Cosmology , class Octree , class Type , class Parameter > void Catalogues::relCat_with_previous_cat (const Point &vobs, std::string &filename, const Point &observer, std::vector< std::array< double, 18 > > &previous_catalogue, const Parameter &parameters, const Cosmology &cosmology, const Octree &octree, const Typelength, const Typeh)\fC [static]\fP"

.PP
Give new position of points\&. Give lensed position of points given points\&. 
.PP
\fBTemplate Parameters:\fP
.RS 4
\fIPoint\fP point type\&. 
.br
\fICosmology\fP Cosmology evolution type\&. 
.br
\fIOctree\fP Octree type\&. 
.br
\fIType\fP Scalar type\&. 
.br
\fIParameter\fP Parameter type\&. 
.RE
.PP
\fBParameters:\fP
.RS 4
\fIvobs\fP Observer peculiar velocity, in SI 
.br
\fIfilename\fP output file name\&. 
.br
\fIobserver\fP Observer position\&. 
.br
\fIprevious_catalogue\fP Previously computed catalogue, used to re-run rejected sources\&. 
.br
\fIparameters\fP Parameter structure 
.br
\fIcosmology\fP Cosmology evolution\&. 
.br
\fIoctree\fP Octree\&. 
.br
\fIlength\fP Spatial length in SI units\&. 
.br
\fIh\fP Dimensionless Hubble parameter 
.RE
.PP


.SH "Author"
.PP 
Generated automatically by Doxygen for MAGRATHEA/PATHFINDER from the source code\&.


src/catalogues.cpp
/* ******************************* CATALOGUES ******************************** */
/*////////////////////////////////////////////////////////////////////////////*/
// PROJECT :        PATHFINDER
// TITLE :          Catalogues
// DESCRIPTION :    Main function of the raytracer to create catalogues
// AUTHOR(S) :      Vincent Reverdy (vince.rev@gmail.com), Michel-Andrès Breton (michel-andres.breton@obspm.fr)
// CONTRIBUTIONS :  [Vincent Reverdy (2012-2013), Michel-Andrès Breton (2015-2021)]
// LICENSE :        CECILL-B License
/*////////////////////////////////////////////////////////////////////////////*/
/// \file           catalogues.cpp
/// \brief          Main function of the raytracer to create catalogues
/// \author         Vincent Reverdy (vince.rev@gmail.com), Michel-Andrès Breton (michel-andres.breton@obspm.fr)
/// \date           2012-2013, 2015-2021
/// \copyright      CECILL-B License
/*////////////////////////////////////////////////////////////////////////////*/



// ------------------------------ PREPROCESSOR ------------------------------ //
#include <ctime>
// Include C++
#include <iostream>
#include <iomanip>
#include <fstream>
#include <algorithm>
#include <utility>
#include <random>
#include <string>
#include <vector>
#include <deque>
#include <array>
#include <tuple>
#include <atomic>
#include <mutex>
// Include libs
#include <mpi.h>
// Include project
#include "magrathea/simplehyperoctree.h"
#include "magrathea/simplehyperoctreeindex.h"
#include "magrathea/hypersphere.h"
#include "magrathea/hypercube.h"
#include "magrathea/constants.h"
#include "magrathea/evolution.h"
#include "magrathea/timer.h"
#include "cone.h"
#include "utility.h"
#include "input.h"
#include "output.h"
#include "integrator.h"
#include "miscellaneous.h"
#include "catalogues.h"
// Octree
#ifdef VELOCITYFIELD
#include "gravity2.h" // with velocity slots
#else
#include "gravity.h"
#endif
// Include C++ HDF5 project
#include "TReadHDF5.h"


// In miscellaneous.h we use std::sample. For versions of gcc lower than 7.x, please use std::experimental::sample. This can be done in the Makefile

// Misc
using namespace magrathea;
// -------------------------------------------------------------------------- //



// ---------------------------------- MAIN ---------------------------------- //
// Main function
/// \brief          Main function.
/// \details        Main function of the program. It takes a namelist as a 
///                 parameter and run the raytracing using MPI parallelization.
///                 To compile it, use make
///                 To run it, use:
///                 mpirun -np NTASKS ./raytracer raytracer.txt
/// \param[in]      argc Number of arguments.
/// \param[in]      argv List of arguments.
/// \return         Zero on success, error code otherwise.
int main(int argc, char* argv[])
{
    // Constants

    using integer = int;
    using uint = unsigned int;
    using real = double;
    using floating = float;
    using point = std::array<real, 3>;
    using position = std::ratio<0>;
    using extent = std::ratio<EXTENT>;
    using indexing = __uint128_t;
    static constexpr uint zero = 0;
    static constexpr uint one = 1;
    static constexpr uint two = 2;
    static constexpr uint dimension = 3;
    static constexpr uint nreference = 5; // Used to set homogeneous octree
    static constexpr real rposition = static_cast<real>(position::num)/static_cast<real>(position::den);
    static constexpr point center({{rposition, rposition, rposition}});
    static constexpr real diameter = static_cast<real>(extent::num)/static_cast<real>(extent::den);
    static const std::string all = "all";
    static const std::string outputsep = "_"; // (Separator used in file names)
    static const std::string outputint = "%05d"; // (Integer format in file names)
    static const std::string outputopening = "%8.6f"; // (Opening angle format in file names)
    static const std::string namelist = argc > 1 ? std::string(argv[1]) : std::string("raytracer.txt"); 

    // Parameters
    std::map<std::string, std::string> parameter;
    integer nthreads = std::thread::hardware_concurrency();
    integer ntasks = nthreads*zero;
    integer rank = zero;

    // Message passing interface
    MPI_Init(&argc, &argv);
    MPI_Comm_size(MPI_COMM_WORLD, &ntasks);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    // Read parameter file
    Miscellaneous::TicketizeFunction(rank, ntasks, [=, &parameter]{parameter = Input::parse(namelist);});
    // Convert strings and put it in struct
    Catalogues::ReadParamFile(parameters, parameter);
    // Initialization
    FileList conefile(parameters.conefmt, zero, parameters.ncones, zero, parameters.conedir);
    SimpleHyperOctree<real, SimpleHyperOctreeIndex<indexing, dimension>, Gravity<floating, dimension>, dimension, position, extent> octree;
    HyperSphere<dimension, point> sphere(center, diameter/two);
    std::vector<std::string> filelist;
    std::vector<Cone<point> > cone(parameters.ncones);
    std::vector<Cone<point> > coneIfRot(parameters.ncones);
    std::array<std::vector<real>, two+two> cosmology;
    Evolution<Photon<real, dimension> > reference;
    real h = zero;
    real omegam = zero;
    real lboxmpch = zero;
    std::ofstream stream;
    std::string filename;
    real thetay(0), thetaz(0);
    std::array< std::array< double, 3 >, 3 > rotm1 = {{zero}};
    const point vobs = {parameters.v0x, parameters.v0y, parameters.v0z};
    const point vobs0 = {0,0,0}; // No peculiar velocity for homogeneous quantities

    if(rank == 0) 
	std::cout<<"#### MAGRATHEA_PATHFINDER "<<std::endl; 
    // Generate cones
    if(parameters.use_previous_catalogues == 0){ // Only useful to assign sources to cone
        if(parameters.isfullsky){
	    if(rank == 0)
	        std::cout<<"## Fullsky cone"<<std::endl;
            Miscellaneous::GenerateFullskyCones(parameters.ncones, cone, coneIfRot, sphere);
        } else{
	    if(rank == 0)
	        std::cout<<"## Narrow cone"<<std::endl;
            Miscellaneous::GenerateNarrowCones(parameters, cone, coneIfRot, sphere, rotm1, thetay, thetaz);
        }
    }
    // Read cosmology
    cosmology = Input::acquire(parameters, h, omegam, lboxmpch);
    if(rank == 0){
        std::cout<<"## Parameter file "<<parameters.paramfile<<std::endl;
        std::cout<<"## Evolution file "<<parameters.evolfile<<std::endl;
        std::cout<<"## Cosmology : h = "<<h<<", omega_m = "<<omegam<<", rhoch2 = "<<parameters.rhoch2<<std::endl;
        std::cout<<"## Box length : "<<lboxmpch<<" Mpc/h, 1 Mpc = "<<parameters.mpc<<" meters "<<std::endl;
        std::cout<<"## Coarse level : "<<parameters.ncoarse<<", number of cones = "<<parameters.ncones<<std::endl;
    }
    const double length = lboxmpch*parameters.mpc/h;

    // Construct homogeneous tree 
    Input::homogenize(octree.assign(nreference, zero));
    reference.append(Integrator::launch(center[zero], center[one], center[two], center[zero]+diameter/two, center[one], center[two]));
    // Propagate a photon in a homogeneous cosmology
    Integrator::integrate<-1>(reference, cosmology, octree, vobs0, length, EXTENT*std::pow(two, static_cast<uint>(std::log2(std::get<0>(cosmology).size()/std::pow(two, nreference)+one)+one)+one));
    cosmology = Input::correct(cosmology, reference);
    octree.fullclear();

    // Execution
#ifndef VELOCITYFIELD
    if(rank == 0) 
	std::cout<<"# Compute relativistic catalogs"<<std::endl;
    reference.fullclear();
    for (uint icone = parameters.firstcone; icone < parameters.lastcone + one; ++icone) {
        if (icone%static_cast<uint>(ntasks) == static_cast<uint>(rank)) {
	    std::vector < std::array< double, 18 > > previous_catalogue;
	    if(parameters.use_previous_catalogues){
                Miscellaneous::ReadFromCat(icone, parameters, previous_catalogue); // For 'rejected', read the id here, and choose only those that are inside the cone
#ifdef VERBOSE
		std::cout<<"# Rank "<<rank<<" cone "<<icone<<" number of sources in the cone : "<<previous_catalogue.size()<<std::endl;
#endif
		if(previous_catalogue.size() == 0){
		    std::cout<<"# Rank "<<rank<<" cone "<<icone<<" no data, skip this cone"<<std::endl;
		    continue;
		}
	    }
	    // Load octree
	    Miscellaneous::loadOctree(icone, octree, conefile);
	    // Set observer position
	    point observer; 
	    observer[0] = 0;
	    observer[1] = 0;
	    observer[2] = 0;

            if(parameters.use_previous_catalogues == 1){ // Rerun previously computed catalogs
		// Set filename
		std::string conetype = parameters.isfullsky? "fullsky" : "narrow";
		std::string sourcetype = parameters.halos ? "halos" : "part";
		std::string jacobinfo = (parameters.jacobiantype == "bundle") ? Output::name(parameters.stop_bundle, outputsep, parameters.plane, outputsep, std::make_pair(outputopening, parameters.openingmin)) : parameters.jacobiantype;
		filename = Output::name(parameters.outputdir, parameters.outputprefix, outputsep, conetype, outputsep, jacobinfo, outputsep, sourcetype, outputsep, std::make_pair(outputint, icone));
		// Re-run sources in already computed catalogue
		Catalogues::relCat_with_previous_cat(vobs, filename, observer, previous_catalogue, parameters, cosmology, octree, length, h);
		     
	    } else if(parameters.use_previous_catalogues == 0 || parameters.use_previous_catalogues == 2){ // If need to compute catalogue or re-run non-converged sources

         	std::vector< std::array< double, 8 > > caractVect_source;
		// Read source files (usually depends on some convention in the filenames)
		if(parameters.typefile == 1){
		    Catalogues::ReadParticlesHDF5(rank, parameters, caractVect_source); 
		} else if(parameters.typefile == 2){
		    Catalogues::ReadParticlesASCII(rank, parameters, caractVect_source); 			
		} else{
		    if(rank == 0){
			std::cout<<"Targets are only available with the HDF5 or ASCII format"<<std::endl;
			std::cout<<"# Error at file "<<__FILE__<<", line : "<<__LINE__<<std::endl;
			std::terminate();
		    }
		}
 
		std::vector< std::array< double, 8 > > targets_position;
		// Assign sources to cone
		targets_position = Miscellaneous::getTargets(caractVect_source, cone[icone], cone);
#ifdef VERBOSE
		std::cout<<"# Rank "<<rank<<" cone "<<icone<<" HALOS : number of sources in the cone : "<<targets_position.size()<<" of "<<caractVect_source.size()<<std::endl;
#endif
                Miscellaneous::fullclear_vector(caractVect_source);
		// If re-run rejected sources 
		if(parameters.use_previous_catalogues == 2){
		    std::vector< std::array< double, 8 > > targets_position_tmp;
		    // Sort sources from previously computed catalogue with index
                    std::sort(previous_catalogue.begin(), previous_catalogue.end(), [](const std::array< double, 18 >& a, const std::array< double, 18 >& b) {return a[0] < b[0];});
		    // Sort sources from full dataset with index
                    std::sort(targets_position.begin(), targets_position.end(), [](const std::array< double, 8 >& a, const std::array< double, 8 >& b) {return a[6] < b[6];});
		     
		    for(uint i = 0; i < previous_catalogue.size(); i++){
		        std::array< double, 8 > point_tmp;
		        point_tmp[6] = previous_catalogue[i][0];
		        const unsigned long int marked = std::distance(std::begin(targets_position), std::upper_bound(std::begin(targets_position), std::end(targets_position), point_tmp, [](const std::array< double, 8 >& first, const std::array< double, 8 >& second){return first[6] < second[6];}));
			// If we find a source that was rejected, then append the vector to re-run it from the beginning
			targets_position_tmp.push_back(targets_position[marked - (marked > 0)]);
		    }
                    Miscellaneous::fullclear_vector(targets_position);	
		    targets_position = targets_position_tmp; 
                    Miscellaneous::fullclear_vector(targets_position_tmp);	
		}
		// Set filename
		std::string conetype = parameters.isfullsky? "fullsky" : "narrow";  
		std::string sourcetype = parameters.halos ? "halos" : "part";
		std::string jacobinfo = (parameters.jacobiantype == "bundle") ? Output::name(parameters.stop_bundle, outputsep,parameters.plane,outputsep,std::make_pair(outputopening, parameters.openingmin)) : parameters.jacobiantype;
		if(parameters.use_previous_catalogues == 2){
		    filename = Output::name(parameters.outputdir, "/rejected/", parameters.outputprefix, outputsep, conetype, outputsep, jacobinfo, outputsep, sourcetype, outputsep, std::make_pair(outputint, icone));
		} else{
		    filename = Output::name(parameters.outputdir, parameters.outputprefix, outputsep, conetype, outputsep, jacobinfo, outputsep, sourcetype, outputsep, std::make_pair(outputint, icone));
		}
		// Run the root-finding method to connect the observer and sources
		Catalogues::relCat(vobs, rotm1, filename, observer, targets_position, previous_catalogue, parameters, cosmology, octree, length, h);
		    
	    } else{
		std::cout<<"# WARNING : If 'use_previous_catalogues' is non-zero, must be equal to 1 or 2"<<std::endl;
		std::cout<<"# Error at file "<<__FILE__<<", line : "<<__LINE__<<std::endl;
		std::terminate();
	    }
	}
    } // for icone
#else
    std::cout<<"Do no forget to disable the -DVELOCITYFIELD option in the Makefile if you want to compute relativistic catalogues"<<std::endl;
    std::cout<<"# Error at file "<<__FILE__<<", line : "<<__LINE__<<std::endl;
    std::terminate();
#endif
	
    
    // Finalization
    MPI_Barrier(MPI_COMM_WORLD);
    MPI_Finalize();
    if(rank == 0)
        std::cout<<"# Run completed !"<<std::endl;
    return 0;
}
// -------------------------------------------------------------------------- //



/*////////////////////////////////////////////////////////////////////////////*/
